### ISSUE TITLE
"this" undefined

### ISSUE THREAD

**Issue by fcmatteo** [stampit-org/react-stampit#9]
Hi, I am trying to use the library to make an isomorphic application. I have a component like this:

``` javascript
const CreateComponent = stampit(React, {
  state: {
    text: 'something'
  },

  changeState: function(event) {
    console.log(this.state.text);
  },

  render: function() {
    return (
      <input type='button' defaultValue={this.state.text} onClick={this.changeState} />
    );
  }
});
```

It renders, both server and client side, but when I click on it I receive this error `Uncaught TypeError: Cannot read property 'state' of undefined`.

I am rendering server-side like this

``` javascript
var CreateComponent = React.createFactory(require('../app/components/CreateComponent.jsx'));
router.get('/component/create/', function (req, res, next) {

  let createComponentReact = React.renderToString(CreateComponent({}));
  res.render('admin/createComponent',
    { title: 'Title',
      createComponentReact: createComponentReact });
});
```

And client-side like this

``` javascript
var CreateComponent = require('./components/CreateComponent.jsx');

React.render(
  <CreateComponent />,
  document.getElementById('CreateComponent')
);
```

**Comment by troutowicz** [comment]
@fcmatteo Thank you for reporting this unexpected, while correct behavior. The reason this does not work is because of my naive understanding of `this` and what React's autobinding actually did in `.createClass`.

 In JS, function calls are late bound. This means that the meaning of `this` inside a function may not be what you expect after passing the function around. React's autobinding handled this behavior automatically for non-React methods in `.createClass`. To make `this` behave as you would expect, you will need to use the [bind](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind) method for event handlers.

We could implement magical autobinding for react-stampit, but I don't want to. I want the user to be in control, and be aware of how JS works instead of relying on magic to make things cognitively simpler.

I will update the README soon with a more appropriate example.

TL;DR

Use [bind](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind) for event handlers.

**Comment by fcmatteo** [comment]
Thanks, that worked fine! I know this isn't Stack Overflow but I have another problem related to this and I don't know if it's problem of the library or I'm doing something wrong (I'm new with React, and with Stampit too).

I have this:

``` javascript
const CreateComponent = stampit(React, {
  state: {
    data: getStoreState()
  },

  _onChange: function() {
    console.log(this);
  },

  componentDidMount: function() {
    CTStore.addChangeListener(this._onChange);
  },

  componentWillUnmount: function() {
    CTStore.removeChangeListener(this._onChange);
  },

  render: function() {
    return (
      <div>something</div>
    );
  }
});
```

And `this` in the `_onChange` method isn't the CreateComponent, I think it's the scope of the CTStore.
If I do the same with React.createClass I get the component scope so I can change the state and render.

I don't want you to solve my problem, just want to know if it's a library bug or I have some kind of misconception.

**Comment by troutowicz** [comment]
Your latest issue should be solvable the same way as the first one was. :)
